### ISSUE TITLE
Question: how to write configurable "source"

### ISSUE THREAD

**Issue by Fuco1** [flycheck/flycheck#1236]
In modern PHP projects use something called "Autoloading", which is a mechanism that depends on the file names and directory structure to find files and load them into the current runtime.

For this reason we can not use `source` or `source-inplace` because both change the file placement or its name and the autoloading mechanism would fail to work.

Using `source-original` works in many cases, but increasingly people use docker or other containers to work in separate environment (mapping the local files as volumes).  The checker is then invoked inside the container with the environment other than what you have on the host.  Which means the `source-original` token is useless.

Right now I use a hardcoded `(eval (replace-regexp-in-string "^/home/matus/dev/" "/usr/src/" (buffer-file-name)))` which obviously is not something I would want to contribute to flycheck (I am preparing a patch for  [phpstan](https://github.com/phpstan/phpstan) support).

How would you handle the case where:

* I need to pass the original buffer to flycheck when developing locally
* But I need some option to pass the file name transformed in some manner on per-project basis (with a directory-local binding ideally)

**Comment by fmdkdd** [comment]
At the outset, I would say a `defcustom` function that takes the filename before passing it to the checker, giving you the chance to transform it.  By default, no transformation would be done.

We have something similar with `flycheck-executable-find` which I understand is helpful for people running their checkers inside `nix-shell`.

**Comment by Fuco1** [comment]
Okey, I just wanted to make sure this is an option that would get included.  It feels a bit "dirty" but that's how all of PHP is, after all :D

I will try to prepare a pull request shortly.

**Comment by fmdkdd** [comment]
Wait wait, I didn't say it would get included!  Just that it's not too far-fetched considering we already do that for executables.

Let's ask @cpitclaudel first :)

**Comment by Fuco1** [comment]
Oh, right.  Now that I think of it a bit more broadly, some generic way to transform file names in and out of "containerized" environments would probably be handy.

**Comment by cpitclaudel** [comment]
I think I need a bit more details about the use case.  @Fuco1, can you share a concrete example? I feel a bit uneasy munging file names, but I don't think I have a good picture of the requested feature yet.

(Also to keep in mind if we do this: it would be nice if the changes were compatible with potential-hypothetical-maybe future work on tramp support)

**Comment by cpitclaudel** [comment]
As a concrete element explaining why I'm uneasy, I'd much rather open an issue with phpstan to allow input on stdin, for example.  Getting rid of temporary files entirely would be nice.

**Comment by Fuco1** [comment]
@cpitclaudel This is unfortunately impossible in PHP.  As explained above, the language relies on the files, their names and their location in order to find the classes and load them.  You could probably hack around it in some rather arcane ways but that is not fixable on phpstan's end but rather every user in every codebase would have to do something fishy, like setting up their own crawlers and stuff.

Very often people also require files *relative* to the current source, with constructs like `require __DIR__ . '/../../some-file.php'`.  This would stop working at once.

My usecase here is that, since setting up multiple versions of PHP is difficult, we use docker to achieve that.  We mount the current directory as a volume inside the container, but there it has a different "root", and so a file `/home/matus/dev/application/file.php` ends up being (for example) `/srv/app/file.php`.  I need a way to transform my "host" filename (that is what I edit with Emacs) to the internal container name (that is where the checker runs).

Does this make sense?

**Comment by cpitclaudel** [comment]
> Very often people also require files relative to the current source, with constructs like require __DIR__ . '/../../some-file.php'. This would stop working at once.

No, I don't think so.  Most checkers that take a source file on stdin also take the name of that file as a command line argument, for the reasons you mentioned.

> My usecase here is that, since setting up multiple versions of PHP is difficult, we use docker to achieve that. We mount the current directory as a volume inside the container, but there it has a different "root", and so a file /home/matus/dev/application/file.php ends up being (for example) /srv/app/file.php. I need a way to transform my "host" filename (that is what I edit with Emacs) to the internal container name (that is where the checker runs).

Yes, it does make sense.  I think I'm ok with having a function that transforms paths, if you're ok with staying around and helping with bugs that may arise in connection with it :)

**Comment by Fuco1** [comment]
@cpitclaudel Thanks for the feedback.  I have quite a ton of Emacs projects myself, so I can't promise I will be very available, but I'm willing (with a delay) to poke around :)  Maybe we can mark this feature experimental?

Now we have to decide how general this should be.  Because for many people the `source-original` would be just fine.  So there are options of adding a new property to the checker, or instead hard-coding the translation in the checker (via eval and some option defaulting to identity).

**Comment by cpitclaudel** [comment]
Sorry for the delay. Would this have to be in the checker? I was expecting something more like a flycheck-file-name-filter-function that would get set buffer-locally.

**Comment by Fuco1** [comment]
I'm coming back to this as it annoys me way to much I have to set things up for every project :D

> Also to keep in mind if we do this: it would be nice if the changes were compatible with potential-hypothetical-maybe future work on tramp support

There is also a docker-tramp method which can use tramp to poke inside docker containers.  Might be worth thinking about that.  I can't straight up think of any other use-case of the file name filter than containers anyway.

> flycheck-file-name-filter-function

Yes, I think I agree with this.  Probably if one checker needs this sort of thing all of them would.

**Comment by cpitclaudel** [comment]
I think we're on the same page here: it would be quite fine to have that function included.  I'm happy to help with reviewing a PR.  @Fuco1 , are you still interested in working on this?

**Comment by bbatsov** [comment]
Closing this as stale. This is a very specific use case (configurable source paths for Docker-based development) and has been open since 2018 without further activity. If you're still looking for this functionality, consider using directory-local variables or a wrapper script to adjust paths before invoking the checker. Feel free to open a new issue or PR if you have a concrete proposal.
